invenio_vocabularies/contrib/funders/schema.py

```diff
@@ -8,17 +8,95 @@
 
 """Funders schema."""
 
+from functools import partial
+
 from flask_babelex import lazy_gettext as _
-from marshmallow import Schema, fields, validate
-from marshmallow_utils.fields import SanitizedUnicode
+from marshmallow import Schema, ValidationError, fields, post_load, pre_dump, \
+    pre_load, validate, validates_schema
+from marshmallow_utils.fields import IdentifierSet, SanitizedUnicode
+from marshmallow_utils.schemas import IdentifierSchema
+
+from ...services.schema import BaseVocabularySchema
+from .config import funder_schemes
 
 
 class FunderRelationSchema(Schema):
     """Funder schema."""
 
+    name = SanitizedUnicode(
+        validate=validate.Length(min=1, error=_('Name cannot be blank.'))
+    )
+    id = SanitizedUnicode()
+
+    @validates_schema
+    def validate_funder(self, data, **kwargs):
+        """Validates that either id either name are present."""
+        id_ = data.get("id")
+        name = data.get("name")
+        if id_:
+            data = {"id": id_}
+        elif name:
+            data = {"name": name}
+
+        if not id_ and not name:
+            raise ValidationError(
+                _("An existing id or a free text name must be present"),
+                "affiliations"
+            )
+
+
+class FunderSchema(BaseVocabularySchema):
+    """Service schema for funders."""
+
+    pid = SanitizedUnicode(required=True)
     name = SanitizedUnicode(
         required=True,
         validate=validate.Length(min=1, error=_('Name cannot be blank.'))
     )
-    scheme = SanitizedUnicode()
-    identifier = SanitizedUnicode()
+    country = SanitizedUnicode()
+    identifiers = IdentifierSet(fields.Nested(
+        partial(
+            IdentifierSchema,
+            allowed_schemes=funder_schemes,
+            identifier_required=False
+        )
+    ))
+
+    # TODO: the 3 hooks below could be extracted to a PIDField
+    @pre_load(pass_many=False)
+    def load_pid(self, data, **kwargs):
+        """Loads the pid.
+
+        If it does not exist, it tries to extract it from the identifiers.
+        """
+        if data.get("pid"):
+            return data
+
+        identifiers = data.get("identifiers", [])
+        for identifier in identifiers:
+            # FIXME: make scheme configurable
```

Comment:
Needs fixing, maybe with a config var (similar to `SCHEMES` in affiliations)

Reply:
I'm a bit hesitant actually about allowing "flexibility" in this since we might end up with "duplicate" entries (under different identifiers, etc.).

Maybe just keeping the first `if data.get("pid"):` is the safest, and removes complexity, i.e. you have to explicitly pass the `pid` value when you create the record and that's the one and only way.

Reply:
I 100% agree, I added it there because we somehow talked about it. Will remove.

Reply:
As a result, the whole `load_pid` hook gets removed. The first `if` clause performs no changes on the incoming data.